cpp/src/arrow/builder.cc

```diff
@@ -872,8 +894,124 @@ Status BooleanBuilder::Append(const std::vector<bool>& values) {
 // ----------------------------------------------------------------------
 // DictionaryBuilder
 
+using internal::DictionaryScalar;
 using internal::WrappedBinary;
 
+namespace {
+
+// A helper class to manage a hash table embedded in a typed Builder.
+template <typename T, typename Enable = void>
+struct DictionaryHashHelper {};
+
+// DictionaryHashHelper implementation for primitive types
+template <typename T>
+struct DictionaryHashHelper<T,
+                            typename std::enable_if<TypeTraits<T>::is_primitive>::type> {
+  using Builder = typename TypeTraits<T>::BuilderType;
+  using Scalar = typename DictionaryScalar<T>::type;
+
+  // Get the dictionary value at the given builder index
+  static Scalar GetDictionaryValue(const Builder& builder, int64_t index) {
+    const Scalar* data = reinterpret_cast<const Scalar*>(builder.data()->data());
+    return data[index];
+  }
+
+  // Compute the hash of a scalar value
+  static int64_t HashValue(const Scalar& value, int byte_width) {
+    return HashUtil::Hash(&value, sizeof(Scalar), 0);
+  }
+
+  // Return whether the dictionary value at the given builder index is unequal to value
+  static bool SlotDifferent(const Builder& builder, int64_t index, const Scalar& value) {
+    return GetDictionaryValue(builder, index) != value;
+  }
+
+  // Append a value to the builder
+  static Status AppendValue(Builder& builder, const Scalar& value) {
+    return builder.Append(value);
+  }
+
+  // Append another builder's contents to the builder
+  static Status AppendBuilder(Builder& builder, const Builder& source_builder) {
+    return builder.AppendValues(
+        reinterpret_cast<const Scalar*>(source_builder.data()->data()),
+        source_builder.length(), nullptr);
+  }
+};
+
+// DictionaryHashHelper implementation for StringType / BinaryType
+template <typename T>
+struct DictionaryHashHelper<
+    T, typename std::enable_if<TypeTraits<T>::is_binary_like>::type> {
```

Comment:
Can you use `enable_if_binary` here https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/util-internal.h#L71? This may also make the `is_binary_like` type traits unneeded

Reply:
Wow, I didn't know about those predicates :-)

Reply:
Done.